src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1916 @@
+
+        r = job_new(JOB_DESCRIBE, t, msg, target_method_describe_finish, NULL, &j);
+        if (r < 0)
+                return r;
+
+        j->version = strdup(version);
+        if (!j->version)
+                return log_oom();
+
+        j->offline = offline;
+
+        r = job_start(j);
+        if (r < 0)
+                return sd_bus_error_set_errnof(error, r, "Failed to start job: %m");
+        TAKE_PTR(j); /* Avoid job from being killed & freed */
+
+        return 1;
+}
+
+static int target_method_check_new_finish(
+                sd_bus_message *msg,
+                const Job *j,
+                sd_json_variant *json,
+                sd_bus_error *error) {
+        const char *reply;
+
+        assert(json);
+
+        sd_json_variant *v = sd_json_variant_by_key(json, "available");
+        if (!v) {
+                log_sysupdate_bad_json("check-new", "Missing key 'available'");
+                return -EINVAL;
+        }
+
+        if (sd_json_variant_is_null(v))
+                reply = "";
+        else
+                reply = sd_json_variant_string(v);
+        if (!reply)
+                return -EINVAL;
+
+        return sd_bus_reply_method_return(msg, "s", reply);
+}
+
+static int target_method_check_new(sd_bus_message *msg, void *userdata, sd_bus_error *error) {
+        Target *t = ASSERT_PTR(userdata);
+        _cleanup_(job_freep) Job *j = NULL;
+        int r;
+
+        assert(msg);
+
+        const char *details[] = {
+                "class", target_class_to_string(t->class),
+                "name", t->name,
+                "offline", "0",
+                NULL
+        };
+
+        r = bus_verify_polkit_async(
+                        msg,
+                        "org.freedesktop.sysupdate1.check",
+                        details,
+                        &t->manager->polkit_registry,
+                        error);
+        if (r < 0)
+                return r;
+        if (r == 0)
+                return 1; /* Will call us back */
+
+        r = job_new(JOB_CHECK_NEW, t, msg, target_method_check_new_finish, NULL, &j);
+        if (r < 0)
+                return r;
+
+        r = job_start(j);
+        if (r < 0)
+                return sd_bus_error_set_errnof(error, r, "Failed to start job: %m");
+        TAKE_PTR(j); /* Avoid job from being killed & freed */
+
+        return 1;
+}
+
+static int target_method_update_finished_early(
+                sd_bus_message *msg,
+                const Job *j,
+                sd_json_variant *json,
+                sd_bus_error *error) {
+
+        /* Called when job finishes w/ a successful exit code, but before any work begins.
+         * This happens when there is no candidate (i.e. we're already up-to-date), or
+         * specified update is already installed. */
+        sd_bus_error_set_errnof(error, -EALREADY, "Job finished with no work: Already up to date.");
+        return sd_bus_error_set_errno(error, -EALREADY);
```

Comment:
now you are setting it twice. the first one will have no effect.


also, please return proper dbus errors, i.e. make up a proper dbus error string here and return it. i.e. use sd_bus_error_setf()

Reply:
I've used the `sd_bus_error_set_errnof` line, which should set a proper dbus error string.

Reply:
sure, but it does not set the bus error name, and it should.

sd-bus has infra to map posix error numbers ("errno", such as EALREADY) to dbus error names. That's a bit ugly though, since posix error numbers are a small set of pre-defined, cryptic error codes, while dbus errors are a lot more descriptive, and full strings. We only want that the errno-to-bus-error logic is used when we receive such an errno from some low-level call, and we have to turn it into a proper dbus error before we can return it. However, in cases where *we* issue an error we shouldn't involve that logic, but instead return a proper dbus error directly.

* sd_bus_error_setf() is what generates proper dbus errors
* sd_bus_error_set_errnof() is what generates dbus errors from posix error numbers

Hence here, where we synthesize the error ourselves, please use sd_bus_error_setf().

To see how this is done, grep for BUS_ERROR_HOME_BUSY for example.